### DISCUSSION TITLE
Min rewrite.

### DISCUSSION THREAD

**Discussion by Milo123459** [minbrowser/min#1447] [Ideas]
Hello, this is an idea I've had on my head for like, a month or something.

Rewriting min to:
Vue 3
TypeScript
TailwindCSS

Over all, would make the code a lot easier and would reduce the code size greatly.

**Comment by shalva97** [comment]
That will probably take months and lot of testing... I think adding some JSDocs and refactoring would be best, it is called Min Browser so let the code also be minimal!

**Comment by Milo123459** [comment]
Of course it'll take ages.

**Comment by PalmerAL** [comment]
I'm glad you're thinking about this, and typescript is nice! If I was starting over on this project, there's a good chance I would use that. However, everything we choose to do is a balance between the effort invested and how much impact it has. Like Shalva says, rewriting the whole project would be a huge amount of effort, and I don't think the impact would be very high. It wouldn't obviously provide value to users - a rewrite likely wouldn't improve performance, or enable any new features, or make the browser more stable. It would potentially make it easier to contribute to the project, but I think that's an issue that can be mostly addressed with better documentation, more organization, and general refactoring (all of which are significantly easier than a full rewrite).

**Reply by Milo123459** [reply]
I think it'd be better. Easier to contribute = better features sooner.

**Comment by lidingcai** [comment]
Maybe you can see this [https://github.com/lidingcai/real-browser](https://github.com/lidingcai/real-browser)

**Reply by PalmerAL** [reply]
Nice work! A few questions for you:

* How much time approximately did this take you?
* Did you use any automated tools to help with the conversion?
* How close to "done" is this, and how much time do you think it would take to complete adding type annotations if it isn't?

**Comment by lidingcai** [comment]
- How much time approximately did this take you?
I don't know exact time, maybe 14 work days need. (Real 14*10 hours work time)

- Did you use any automated tools to help with the conversion?
I didn't use automated tools to transfer the code, just used a tool to analysised the code to get the code dependence, so I know what files to transfer first. And use VS Code to get the hints.

- How close to "done" is this, and how much time do you think it would take to complete adding type annotations if it isn't?
I know I have done >=90% work , I just removed some features I can not use (Something like 1password?? Because actually I cloned the repository last year Spring, so I forget some details.) I build my typescript code yesternight, and the browser works well.

Anyway, I don't think transfer to typescript it's as hard as you think. Because the source code is not huge yet.

**Reply by PalmerAL** [reply]
That's really nice!

Typescript would be better than what we have, and based on your comment, it seems like it would be less work than what I thought originally to migrate.

It sounds like your branch has probably diverged from the main branch a bit, and there's always new changes being merged into main, so we might need to figure out a way to do this migration incrementally. 

Is this something you're interested in contributing into the main repository?

**Comment by lidingcai** [comment]
Yeah, I rewrite that at Spring, 2024, the code is a little old now.
I have interesting to do this, the problem is the time and ability.(I have free time last year.)
To the new merged changes, I think there is no good way to deal with that. Maybe I can try to transfer the whole project to typescript again once I have time (based on one specific commit), and send you the code so less increment change.

**Reply by PalmerAL** [reply]
It seems like it should be possible to migrate things one file at a time (in return for making the type checking less strict), which seems like the most feasible way to do this: https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html

**Comment by digitalextremist** [comment]
Hello all. New to `min` but starting here at the rewrite thread. I am working with `Vue3` and `Tailwind4` and `Typescript` and `Bun` which is the piece which makes all the rest of the rewrite worth more, imo. If I end up sticking with `min` I would help with this.

My suspicion is that the browser itself will need to be packaged with distributed applications. I would see `min` as a possible way to do that, rather than `nativefier` or `WebApps` support on `Mint Linux`, etc. Truly embracing the browser, and including one we tuned.

I do not yet see what "necessary evils" might exist in `min` and am going to audit all the packages. I am coming from `Brave` personally, and foresee the need to actively block use by compromised browsers, like those driven by LLMs in certain cases.

Curious if someone would be able to give me their current assessment of `min` and truly point out its strengths and weaknesses from an insider perspective on the code. I was very glad to hear that direction the stack needs to go, and seem deeply aligned.

Nice work on what I see so far! Seems promising and exciting... eager to deeply audit this and see where it needs to go.

---

And one aside: I am a [`Zed`](https://zed.dev) enthusiast and originally found `min` wishing there was a complete re-write of a browser, likely in `Rust` ... I am equally pulled in that direction and considering a "totally redone" direction. That is why I include the `Bun` aspect, because if `min` were `Bun` based, though it may not outperform `Rust` ever ever ever, it could at least be at the cutting edge within its own language, and have possibly an order-of-magnitude difference in performance and capabilities, perhaps even removing huge sections of what is currently believed to be necessary. 

It seems we can ask ourselves what really matters better now, compared to 2021. I am asking myself: what will survive until 2040, for example? It seems more likely to use `Bun` and avoid a `Rust` rewrite rather than stop everything and rewrite a browser. It seems appropriate to use a 'web' language for THE most 'web' tool. And since everything in the frontend is catering to the browser anyway, it seems like `min` might be the point where field itself can advance. For example, rather than see `min` as a 'browser' we could see it as the foyer of the desktop operating system, which it is.... from a distributed application standpoint especially. The corollary on the backend side would be the virtualization platform, containing and running images.
